code/src/sixsq/nuvla/server/resources/notification.clj

```diff
@@ -0,0 +1,151 @@
+  "
+TODO: provide namespace documentation.
+
+Notification resource.
+
+ACL
+
+Notifications can only be created by admins. Creator of notification
+should provide resource level ACL accordingly, which for example
+may depend on the type of the notification.
+"
+  (:require
+    [sixsq.nuvla.auth.acl-resource :as a]
+    [sixsq.nuvla.db.impl :as db]
+    [sixsq.nuvla.server.resources.common.crud :as crud]
+    [sixsq.nuvla.server.resources.common.schema :as c]
+    [sixsq.nuvla.server.resources.common.std-crud :as std-crud]
+    [sixsq.nuvla.server.resources.common.utils :as u]
+    [sixsq.nuvla.server.resources.resource-metadata :as md]
+    [sixsq.nuvla.server.resources.spec.notification :as notification]
+    [sixsq.nuvla.server.util.metadata :as gen-md]
+    [sixsq.nuvla.server.util.response :as r]))
+
+
+(def ^:const resource-type (u/ns->type *ns*))
+
+
+(def ^:const collection-type (u/ns->collection-type *ns*))
+
+
+(def collection-acl {:query ["group/nuvla-user"]
+                     :add   ["group/nuvla-admin"]})
+
+(def resource-acl {:owners ["group/nuvla-admin"]})
+
+;;
+;; "Implementations" of multimethod declared in crud namespace
+;;
+
+(def validate-fn (u/create-spec-validation-fn ::notification/schema))
+
+
+(defmethod crud/validate
+  resource-type
+  [resource]
+  (validate-fn resource))
+
+
+(def add-impl (std-crud/add-fn resource-type collection-acl resource-type))
+
+
+(defmethod crud/add resource-type
+  [request]
+  (add-impl request))
+
+
+(def retrieve-impl (std-crud/retrieve-fn resource-type))
+
+
+(defmethod crud/retrieve resource-type
+  [request]
+  (retrieve-impl request))
+
+
+(def delete-impl (std-crud/delete-fn resource-type))
+
+
+(defmethod crud/delete resource-type
+  [request]
+  (delete-impl request))
+
+
+;;
+;; use default ACL method
+;;
+
+(defmethod crud/add-acl resource-type
+  [resource request]
+  (if (u/is-collection? resource-type)
+    (assoc resource :acl collection-acl)
+    (assoc resource :acl (merge resource-acl (-> request :body :acl)))))
+
+
+;;
+;; available collection and resource operations
+;;
+
+
+(defn set-collection-ops
+  [{:keys [id] :as resource} request]
+  (if (a/can-add? resource request)
+    (let [ops [{:rel (:add c/action-uri) :href id}]]
+      (assoc resource :operations ops))
+    (dissoc resource :operations)))
+
+
+(defn set-resource-ops
+  [{:keys [id] :as resource} request]
+  (let [can-manage? (a/can-manage? resource request)
+        ops (cond-> []
+                    can-manage? (conj {:rel (:delete c/action-uri) :href id})
+                    can-manage? (conj {:rel (:defer c/action-uri) :href (str id "/defer")}))]
+    (if (seq ops)
+      (assoc resource :operations ops)
+      (dissoc resource :operations))))
+
+
+(defmethod crud/set-operations resource-type
+  [resource request]
+  (if (u/is-collection? resource-type)
+    (set-collection-ops resource request)
+    (set-resource-ops resource request)))
+
+
+(defmethod crud/do-action [resource-type "defer"]
```

Comment:
I think that it would be better if this action took a positive integer as an argument. This could then be added to the current time to provide a timestamp that is known to be valid. It also relieves the client from having to do the formatting on its side. Also, I think that there should be a default value. This allows the operation to work, although with less flexibility. I'd suggest that the argument be named after the units used: e.g. if the value is in minutes then "minutes" would be a good name.  I don't know whether minutes, hours, or days would be a good unit to use.

Reply:
OK. Agreed.